### PR TITLE
[ws-manager-mk2] redact log

### PR DESCRIPTION
**Pull request by iQQBot**
## Description

This PR introduces a new function `DeepCopyStruct` in `scrubber`, it can copy a struct and do scrub.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at dbc828e</samp>

This pull request enhances the logging and scrubbing features of the `ws-manager-mk2` component. It adds a new feature to the `scrubber` package to deep copy and scrub struct values, and uses it to log values securely with the `logrusr` package. It also refactors the logging logic in the workspace controller and removes some unused code.

</details>

## Related Issue(s)

Fixes ENG-872

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - pd-ws-man-5c0d6febc1</li>
	<li><b>🔗 URL</b> - <a href="https://pd-ws-man-5c0d6febc1.preview.gitpod-dev.com/workspaces" target="_blank">pd-ws-man-5c0d6febc1.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - pd-ws-man-2-redact-log-gha.18376</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-pd-ws-man-5c0d6febc1%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by WVerlaek** [comment]
Thanks - do you have examples of how e.g. the workspace status/pod status fields show now in the logs? Which fields are included and which are redacted?

File: components/ws-manager-mk2/main.go [hunk 1]
```diff
@@ -77,7 +79,12 @@ func main() {
 	flag.Parse()
 
 	log.Init(ServiceName, Version, jsonLog, verbose)
-	baseLogger := logrusr.New(log.Log)
+
+	l := log.WithFields(logrus.Fields{})
+	l.Logger.SetReportCaller(false)
+	baseLogger := logrusr.New(l, logrusr.WithFormatter(func(i interface{}) interface{} {
+		return &scrubber.TrustedValueWrap{Value: scrubber.Default.DeepCopyStruct(i)}
```
Comment (easyCZ) [inline components/ws-manager-mk2/main.go:86]: What is the perf implication of doing a deep copy? If I remember correctly, @akosyakov was running into perf issues when we performed deep scrubbing and the call we made is we'd only scrub a fixed depth and redact anything below.

File: components/scrubber/scrubber.go [hunk 2]
```diff
@@ -86,6 +97,8 @@ type Scrubber interface {
 	//   }
 	//
 	Struct(val any) error
+
+	DeepCopyStruct(val any) any
```
Comment (akosyakov) [inline components/scrubber/scrubber.go:101]: Could you add docs explaining the different to Struct? Does it create a new object while scrubbing? What are expectations about interfaces?

File: components/ws-manager-mk2/main.go (hunk 1 above)
Comment (akosyakov) [reply components/ws-manager-mk2/main.go:86]: On complex and deep structures it was considerably slow, but in golang is still was negligible.
Our concern though was that we introduce it generically and don't want to make some hot path even slower by mistake, so decided to go with enforcing rather flat primitive properties. If this code is not on such path I would think it should be alright.

File: components/scrubber/scrubber.go [hunk 3]
```diff
@@ -48,6 +49,16 @@ type TrustedValue interface {
 	IsTrustedValue()
 }
 
+type TrustedValueWrap struct {
```
Comment (akosyakov) [inline components/scrubber/scrubber.go:52]: Why TrustedValue does not fit here? Could you add docs why we need this type and when to use which?

Comment (akosyakov) [reply components/scrubber/scrubber.go:52]: Or can we change TrustedValue instead?

Comment (akosyakov) [reply components/scrubber/scrubber.go:52]: yes, I mean it is rather a library it would be good to have a short comment there for others.

**Comment by akosyakov** [comment]
> actually, this url contains workspace id, but the scriber don't have rule to redact it

@iQQBot I'm not sure it is good idea to scrub each URL string, since regex is very expensive. If you know the structure maybe just scrub this property specifically on the top?

File: components/ws-manager-mk2/main.go (hunk 1 above)
Comment (akosyakov) [reply components/ws-manager-mk2/main.go]: I mean the point is that delivery of updates now will be slowed down by deep scrubbing of workspace status, you need to decide how acceptable it is.

**Review by akosyakov — APPROVED** [review]
approve to unblock, fine with scrubber changes

**Comment by WVerlaek** [comment]
@iQQBot changes lgtm but still see the URL unscrubbed, do you want to do that in a follow-up pr?

**Comment by akosyakov** [comment]
Could you clarify which URLs? Also do you really want to log the entire workspace status? e.g. are you interested in ports?

**Comment by akosyakov** [comment]
for status.url my understanding was
- do deep scrubbing get a new structure
- now get a value from status.url and replace with `[redacted]`?

**Comment by WVerlaek** [comment]
> Also do you really want to log the entire workspace status? e.g. are you interested in ports?

if possible yes - it's been useful to debug workspace issues by having the full status. Lets you reconstruct a timeline of the workspace and its status changes

**Comment by WVerlaek** [comment]
Is it possible to add a `scrubbed` tag on the status.URL field?

**Review by WVerlaek — APPROVED** [review]
lgtm

**Pull request merged**